### ISSUE TITLE
Could self-certainty use k3 estimator to estimate like which is in GRPO?

### ISSUE THREAD

**Issue by Psychim** [sunblaze-ucb/Intuitor#14]
Considering sum over the vocabullary could be expensive, is it possible to use k3 estimator instead?

**Comment by XuandongZhao** [comment]
The core calculation is similar to entropy. I believe the computational cost is minimal since the reward is only calculated once, after the generation is complete.

```python
def entropy_from_logits(logits: torch.Tensor):
    """Calculate entropy from logits."""
    pd = torch.nn.functional.softmax(logits, dim=-1)
    entropy = torch.logsumexp(logits, dim=-1) - torch.sum(pd * logits, dim=-1)
    return entropy


def self_certainty_from_logits(logits: torch.Tensor):
    """Calculate self-certainty from logits."""
    self_certainty = torch.logsumexp(logits, dim=-1) - logits.mean(dim=-1)
    return self_certainty
```

**Comment by Psychim** [comment]
Solved with some code optimization. Considering we all do log_softmax in any kind of training which computes things in similar way, it may be not the primary reason.
